libs/domain/feature-loaders/src/discord/DiscordLoader.ts

```diff
@@ -0,0 +1,156 @@
+import { notEmpty } from "@banklessdao/util-misc";
+import {
+    LoadContext,
+    ScheduleMode,
+    DataLoaderBase,
+    DEFAULT_CURSOR,
+    DatabaseError,
+} from "@shared/util-loaders";
+import {
+    Data,
+    NonEmptyProperty,
+    OptionalProperty,
+} from "@banklessdao/util-schema";
+import * as t from "io-ts";
+import { withMessage } from "io-ts-types";
+import { TextChannel, Client, Intents } from "discord.js";
+import * as TE from "fp-ts/lib/TaskEither";
+
+const INFO = {
+    namespace: "discord",
+    name: "Message",
+    version: "V1",
+};
+
+@Data({
+    info: INFO,
+})
+class DiscordMessage {
+    @NonEmptyProperty()
+    id: string;
+    @OptionalProperty()
+    content?: string;
+    @NonEmptyProperty()
+    createdAt: string;
+}
+
+const DiscordMessageCodec = t.strict({
+    id: t.string,
+    content: t.string,
+    createdAt: t.string,
+});
+
+const DiscordMessagesCodec = t.strict({
+    messages: withMessage(
+        t.array(DiscordMessageCodec),
+        () => "DiscordMessage is required"
+    ),
+});
+
+type DiscordMessages = t.TypeOf<typeof DiscordMessagesCodec>;
+
+export class DiscordLoader extends DataLoaderBase<
+    DiscordMessages,
+    DiscordMessage
+> {
+    public info = INFO;
+
+    // Discord doesn't allow 1000 here
+    protected batchSize = 100;
+    protected type = DiscordMessage;
+    protected cadenceConfig = {
+        [ScheduleMode.BACKFILL]: { seconds: 5 },
+        [ScheduleMode.INCREMENTAL]: { minutes: 5 },
+    };
+
+    protected codec = DiscordMessagesCodec;
+
+    private token: string;
+    private channelId: string;
+
+    client = new Client({
+        intents: [
+            Intents.FLAGS.GUILDS,
+            Intents.FLAGS.DIRECT_MESSAGES,
+            Intents.FLAGS.GUILD_MESSAGES,
+        ],
+    });
+
+    constructor(token: string, channelId: string) {
+        super();
+        this.token = token;
+        this.channelId = channelId;
+
+        // Create a new client instance
+        // Login to Discord with your client's token
+        // TODO we could wait for the ready - but it should be ready quickly, most probably before the loader
+        // TODO in general - could add some discord error handling
+        this.client.login(token);
+    }
+
+    // TODO create a new loader base? -> check "await" TODO below as well
+    protected loadRaw(context: LoadContext) {
+        return TE.tryCatch(
+            this.mapDiscordResult(context.cursor),
+            (err: unknown) => {
+                return new DatabaseError(String(err));
+            }
+        );
+    }
+
+    protected mapDiscordResult(cursor?: string) {
+        const channel = this.client.channels.cache.get(
+            this.channelId
+        ) as TextChannel;
+
+        // TODO do we need to await this here at all? 
```

Comment:
We don't. In `loadRaw` we just load the "raw" result (whatever the API is returning). This is how `load` looks like:

```ts
public load(context: LoadContext) {
    return pipe(
        TE.Do,
        TE.bind("loadContext", () => this.preLoad(context)),
        TE.bind("rawResult", ({ loadContext }) =>
            this.loadRaw(loadContext)
        ),
        TE.bindW("mappedResult", ({ rawResult }) => {
            return TE.tryCatch(
                async () => this.mapResult(rawResult),
                (e) => new UnknownError(e)
            );
        }),
        TE.map((params) => {
            return {
                cursor: this.getNextCursor(params),
                data: params.mappedResult,
            };
        })
    );
}
```

The relevant part starts with the `rawResult` binding: this just executes `loadRaw` and returns whatever it returns, and then constructs the `mappedResult` and it also comes with built-in error-handling. You can use that or add your custom one, it doesn't really matter.

So in your case `mapDiscordResult` should probably go into `loadRaw` and you can just call `channel.messages.fetch` there and return whatever is in the response.

Reply:
refactored codec out to CodecDataLoaderBase